dppl/oneapi_interface.pyx

```diff
@@ -146,12 +168,13 @@ def device_context (dev=device_type.gpu, device_num=0):
     # If set_context is unable to create a new context an exception is raised.
     try:
         ctxt = None
-        ctxt = runtime._activate_queue(dev, device_num)
+        _tls._in_dppl_ctxt = True
+        ctxt = _tls._runtime._activate_queue(dev, device_num)
         yield ctxt
     finally:
         # Code to release resource
         if ctxt:
-            print("Removing the context from the deque of active contexts")
-            runtime._deactivate_current_queue()
+            _tls._runtime._deactivate_current_queue()
+            _tls._in_dppl_ctxt = False
```

Comment:
Shouldn't this be a virtual stack of contexts?  I would think that on "finally" that you pop the top off the stack and then only set in_dppl_ctxt to false if the stack is empty.  Even better, you don't have this in_dppl_ctxt at all and just check the size of the stack if you need to know if you're in a context or not.

Reply:
The stack of contexts is maintained inside the _runtime.rt object. I am using a std::deque to store the list of currently activated queues. The first entry of the deque is always the global or default queue. Using an example:

```
default_q = dppl.get_current_queue()

with dppl.device_context(dppl.device_type.cpu, 0) as cpu_q:
    ...
    with dppl.device_context(dppl.device_type.gpu, 0) as gpu_0_q:
        ...
    with dppl.device_context(dppl.device_type.gpu, 1) as gpu_1_q:
        ...
```
Here the stack will be maintained inside dppl._tls._runtime as:

`default_q` (initial state)
`cpu_q --> default_q` (on entering first with context)
`gpu_0_q --> cpu_q --> default_q` (on entering the first nested with context)
`cpu_q --> default_q` (on exiting the first nested with context)
`gpu_1_q --> cpu_q --> default_q` (on entering the second nested with context)
`cpu_q --> default_q` (on exiting the second nested with context)
`default_q` (on exiting the top-level with context)

The `_deactivate_current_queue()` call does the pop from the `deque. dppl.get_current_queue` always returns the current top of stack queue.

About your second point about ` in_dppl_ctxt`, yes the way it is currently implemented it returns wrong result when contexts are nested. Let me fix it to return the size of the deque -1 (since the default queue is always going to be inside the deque).

Reply:
Made the change in 8318d44